source/taskmanager-client/src/main/java/nl/aerius/taskmanager/client/TaskManagerClientSender.java

```diff
@@ -68,142 +60,102 @@ public TaskManagerClientSender(final BrokerConnectionFactory factory) {
-          final String replyQueueName = startConsumer(wrapper);
-          builder.replyTo(replyQueueName);
-        }
-        final BasicProperties props = builder.build();
-        // Send the message to the taskmanager.
-        channel.basicPublish("", queueName, props, QueueHelper.objectToBytes(task));
+        checkChannel(channel);
+        final BasicProperties.Builder propertiesBuilder = new BasicProperties.Builder();
+        prepareBeforeSend(propertiesBuilder);
+        propertiesBuilder.correlationId(correlationId);
+        propertiesBuilder.messageId(messageId);
+        propertiesBuilder.deliveryMode(DELIVERY_MODE_PERSISTENT);
+
+        channel.get().basicPublish("", queueName, propertiesBuilder.build(), QueueHelper.objectToBytes(data));
         // task has been send successfully, return.
         done = true;
-      } catch (final ConnectException | TimeoutException e) {
+      } catch (final ConnectException e) {
         // don't catch all IOExceptions, just ConnectExceptions.
         // exceptions like wrong host-name should cause a bigger disturbance.
         // those indicate that connection has temporarily been lost with RabbitMQ, though could be not that temporarily...
-        LOG.error("Sending task (id: {}) failed, retrying in a bit.", wrapper.getTaskId(), e);
+        LOG.error("Sending task (id: {}) failed, retrying in a bit.", correlationId, e);
         try {
           Thread.sleep(CLIENT_START_RETRY_PERIOD);
         } catch (final InterruptedException e1) {
           // no need to log.
           Thread.currentThread().interrupt();
         }
-        LOG.info("Retrying to send task (id: {})", wrapper.getTaskId());
+        LOG.info("Retrying to send task (id: {})", correlationId);
       }
     }
   }
 
-  private Connection getConnection() {
-    return factory.getConnection();
-  }
-
   /**
-   * Starts a consumer, but only if a callback is specified.
-   * @param taskWrapper task data
-   * @return name of the reply queue
+   * Method to enrich the builder with additional data before sending the task.
+   *
+   * @param builder builder to enrich
    * @throws IOException
    */
-  private String startConsumer(final TaskWrapper taskWrapper) throws IOException {
-    final Channel channel = getConnection().createChannel();
-    final String replyQueueName = channel.queueDeclare().getQueue();
-    final TaskResultConsumer resultConsumer = new TaskResultConsumer(channel, taskWrapper, this);
-    channel.basicConsume(replyQueueName, true, resultConsumer);
-    return replyQueueName;
+  protected void prepareBeforeSend(final Builder builder) throws IOException {
+    // Default implementation does nothing.
+  }
+
+  protected boolean checkChannel(final AtomicReference<Channel> channelReference) throws IOException {
+    final Channel channel = channelReference.get();
+    if (channel == null || !channel.isOpen()) {
+      channelReference.set(getConnection().createChannel());
```

Comment:
I guess technically there's a chance that another thread will create a 2nd channel between getting the (null) reference and creating-and-setting the new one? Might need a synchronisation on the reference?